python/ql/lib/semmle/python/security/dataflow/NoSQLInjectionQuery.qll

```diff
@@ -0,0 +1,60 @@
+/**
+ * Provides a taint-tracking configuration for detecting NoSQL injection vulnerabilities
+ */
+
+import python
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.Concepts
+private import NoSQLInjectionCustomizations::NoSqlInjection as C
+
+/**
+ * A taint-tracking configuration for detecting NoSQL injection vulnerabilities.
+ */
+module NoSQLInjectionConfig implements DataFlow::StateConfigSig {
+  class FlowState = C::FlowState;
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    source instanceof C::StringSource and
+    state instanceof C::StringInput
+    or
+    source instanceof C::DictSource and
+    state instanceof C::DictInput
+  }
+
+  predicate isSink(DataFlow::Node source, FlowState state) {
+    source instanceof C::StringSink and
+    (
+      state instanceof C::StringInput
+      or
+      // since dictionaries can encode strings
+      state instanceof C::DictInput
+    )
+    or
+    source instanceof C::DictSink and
+    state instanceof C::DictInput
+  }
+
+  predicate isBarrier(DataFlow::Node node, FlowState state) {
+    // Block `StringInput` paths here, since they change state to `DictInput`
+    exists(C::StringToDictConversion c | node = c.getOutput()) and
+    state instanceof C::StringInput
+  }
```

Comment:
That is a fair point. A reformulation of the modelling might indeed make things simpler..

Reply:
I changed the names of the flow states and the comments. I did not actually change the above code, in the end:
- I believe we still need the second case in `isSink`, since we may have a path that starts in a `DictSink` and ends in a `StringSink` without going through any conversion. Your suggestion would remove that path.
- I believe the barrier, while possibly semantically unnecessary, makes the graph lighter and thus more performant.